pkg/helm/urlfetcher.go

```diff
@@ -0,0 +1,165 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package helm
+
+import (
+	"bytes"
+	"crypto/sha256"
+	"encoding/hex"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/mholt/archiver"
+
+	"istio.io/pkg/log"
+)
+
+// FileDownloader is wrapper of HTTP client to download files
+type FileDownloader struct {
+	// client is a HTTP/HTTPS client.
+	client *http.Client
+}
+
+// URLFetcher is used to fetch and manipulate charts from remote url
+type URLFetcher struct {
+	// url is url to download the charts
+	url string
+	// verifyURL is url to download the verification file
+	verifyURL string
+	// verify indicates whether the downloaded tar should be verified
+	verify bool
+	// destDir is path of charts downloaded to, empty as default to temp dir
+	destDir string
+	// downloader downloads files from remote url
+	downloader *FileDownloader
+}
+
+// NewURLFetcher creates an URLFetcher pointing to urls and destination
+func NewURLFetcher(repoURL string, destDir string) *URLFetcher {
+	uf := &URLFetcher{
+		url:        repoURL + "/" + InstallationChartsFileName,
+		verifyURL:  repoURL + "/" + InstallationShaFileName,
+		verify:     true,
+		destDir:    destDir,
```

Comment:
We want this to default to a platform specific temp dir - however, not one that's randomly generated, since this will lead to downloading the same package multiple times.

Reply:
If we run the fetcher by poller, during the poller running time, the destDir does not change, which is under platform default temp dir. Do you mean that even across different poller the destDir name should be fixed?

Reply:
For a poller we are ok, since the process stays alive. The problem is the non-poll mode - if a user runs a CLI command multiple times that targets URL, each time the command will download a fresh copy of the same install package.

Reply:
Actually the NewURLFetcher accept destDir passed in, so the caller can specify whether it is a random path or a fix path. Right now the poller would pass in a random path. To make it more clear, add a check in NewURLFetcher to use a fix path under platform specific temp dir when destDir passed in is empty.